core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/AbstractGroupNode.java

```diff
@@ -22,35 +22,47 @@
 import java.util.Map;
 import java.util.Queue;
 import java.util.function.Consumer;
+import java.util.function.Function;
 import java.util.function.Supplier;
 
-public abstract class AbstractGroupNode<InTuple_ extends Tuple, OutTuple_ extends Tuple, GroupKey_, ResultContainer_>
+public abstract class AbstractGroupNode<InTuple_ extends Tuple, OutTuple_ extends Tuple, GroupKey_, ResultContainer_, Result_>
         extends AbstractNode {
 
     private final int groupStoreIndex;
     private final Supplier<ResultContainer_> supplier;
+    private final Function<ResultContainer_, Result_> finisher;
     /**
      * Some code paths may decide to not supply a collector.
      * In that case, we skip the code path that would attempt to use it.
      */
-    private final boolean runAccumulate;
+    private final boolean hasCollector;
     /**
      * Calls for example {@link AbstractScorer#insert(Tuple)}, and/or ...
      */
     private final Consumer<OutTuple_> nextNodesInsert;
+    /**
+     * Calls for example {@link AbstractScorer#update(Tuple)}, and/or ...
+     */
+    private final Consumer<OutTuple_> nextNodesUpdate;
     /**
      * Calls for example {@link AbstractScorer#retract(Tuple)}, and/or ...
      */
     private final Consumer<OutTuple_> nextNodesRetract;
     private final Map<GroupKey_, Group<OutTuple_, GroupKey_, ResultContainer_>> groupMap;
     private final Queue<Group<OutTuple_, GroupKey_, ResultContainer_>> dirtyGroupQueue;
 
-    protected AbstractGroupNode(int groupStoreIndex, Supplier<ResultContainer_> supplier,
-            Consumer<OutTuple_> nextNodesInsert, Consumer<OutTuple_> nextNodesRetract) {
+    protected AbstractGroupNode(int groupStoreIndex,
+            Supplier<ResultContainer_> supplier,
+            Function<ResultContainer_, Result_> finisher,
+            Consumer<OutTuple_> nextNodesInsert,
+            Consumer<OutTuple_> nextNodesUpdate,
+            Consumer<OutTuple_> nextNodesRetract) {
```

Comment:
Considering how much these three consumers appear *everywhere* and how many repetitious code lines they actually bring, may I suggest creating a new type to encapsulate them?

    interface TupleLifecycle<Tuple_ extends Tuple> {

        void insert(Tuple_ tuple);
        void update(Tuple_ tuple);
        void retract(Tuple_ tuple);

    }

Doesn't need to be done in this PR, but IMO needs to be done. This PR could have been half the diff if we had this already.

Reply:
Agreed, but different issue indeed.

Reply:
See zulip